File: app/src/main/java/com/apps/LauncherCoverLoader.java
package com.apps;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Handler;
import android.os.Looper;
import android.util.LruCache;
import android.widget.ImageView;

import com.yuki.yukihub.util.AppExecutors;

import java.io.InputStream;

/**
 * 封面异步加载器：在 IO 线程解码 + 内存缓存 + 采样降采样，
 * 避免 RecyclerView 绑定期间在主线程同步解码导致的掉帧。
 */
public final class LauncherCoverLoader {
    private LauncherCoverLoader() {}

    /** 内存缓存：取应用可用内存的 1/8，至少 2MB */
    private static final int CACHE_SIZE = Math.max(2 * 1024 * 1024,
            (int) (Runtime.getRuntime().maxMemory() / 8));
    private static final LruCache<String, Bitmap> cache = new LruCache<String, Bitmap>(CACHE_SIZE) {
        @Override
        protected int sizeOf(String key, Bitmap value) {
            return value.getByteCount();
        }
    };

    private static final Handler mainHandler = new Handler(Looper.getMainLooper());

    /** 封面卡片目标尺寸（2 列网格下约 160dp × 144dp，按 3x 密度估算） */
    private static final int TARGET_WIDTH = 480;
    private static final int TARGET_HEIGHT = 432;

    public interface Callback {
        void onLoaded(boolean success);
    }

    public static void loadInto(ImageView imageView, String uriText, Callback callback) {
        if (imageView == null) return;
        final String key = uriText == null ? "" : uriText.trim();
        // 用 tag 记录当前请求，避免回收/复用后旧请求覆盖新内容
        final String prev = (String) imageView.getTag();
        if (key.equals(prev)) return;
        imageView.setTag(key);
        imageView.setImageDrawable(null);

        if (key.isEmpty()) {
            if (callback != null) callback.onLoaded(false);
            return;
        }

        Bitmap cached = cache.get(key);
        if (cached != null) {
            imageView.setImageBitmap(cached);
            if (callback != null) callback.onLoaded(true);
            return;
        }

        final Context ctx = imageView.getContext().getApplicationContext();
        AppExecutors.runOnIo(() -> {
            final Bitmap bitmap = decodeSampled(ctx, key);
            if (bitmap != null) cache.put(key, bitmap);
            mainHandler.post(() -> {
                if (!key.equals(imageView.getTag())) return;
                if (bitmap != null) {
                    imageView.setImageBitmap(bitmap);
                }
                if (callback != null) callback.onLoaded(bitmap != null);
            });
        });
    }

    /** 取消该 ImageView 上的待加载请求并清空显示 */
    public static void clear(ImageView imageView) {
        if (imageView == null) return;
        imageView.setTag(null);
        imageView.setImageDrawable(null);
    }

    private static Bitmap decodeSampled(Context ctx, String uriText) {
        try {
            Uri uri = Uri.parse(uriText);
            BitmapFactory.Options opts = new BitmapFactory.Options();
            opts.inJustDecodeBounds = true;
            try (InputStream is = ctx.getContentResolver().openInputStream(uri)) {
                BitmapFactory.decodeStream(is, null, opts);
            }
            opts.inSampleSize = calculateSampleSize(opts.outWidth, opts.outHeight);
            opts.inJustDecodeBounds = false;
            try (InputStream is = ctx.getContentResolver().openInputStream(uri)) {
                return BitmapFactory.decodeStream(is, null, opts);
            }
        } catch (Throwable ignored) {
            return null;
        }
    }

    private static int calculateSampleSize(int width, int height) {
        if (width <= 0 || height <= 0) return 1;
        int sample = 1;
        while ((width / sample) > TARGET_WIDTH || (height / sample) > TARGET_HEIGHT) {
            sample *= 2;
        }
        return Math.max(1, sample);
    }
}

File: app/src/main/java/com/apps/widget/LauncherCoverLoader.kt
package com.apps.widget

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.util.LruCache
import android.widget.ImageView
import com.core.util.AppExecutors
import com.core.util.RxMainScheduler
import kotlin.math.max

/** Async cover decoder with a bounded in-memory bitmap cache. */
object LauncherCoverLoader {
    private const val TARGET_WIDTH = 480
    private const val TARGET_HEIGHT = 432
    private val cache = object : LruCache<String, Bitmap>(max(2 * 1024 * 1024, (Runtime.getRuntime().maxMemory() / 8).toInt())) {
        override fun sizeOf(key: String, value: Bitmap): Int = value.byteCount
    }

    interface Callback { fun onLoaded(success: Boolean) }

    @JvmStatic
    fun loadInto(imageView: ImageView?, uriText: String?, callback: Callback?) {
        imageView ?: return
        val key = uriText?.trim().orEmpty()
        if (key == imageView.tag as? String) return
        imageView.tag = key
        imageView.setImageDrawable(null)
        if (key.isEmpty()) { callback?.onLoaded(false); return }
        cache[key]?.let { imageView.setImageBitmap(it); callback?.onLoaded(true); return }
        val context = imageView.context.applicationContext
        AppExecutors.runOnIo {
            val bitmap = decodeSampled(context, key)
            bitmap?.let { cache.put(key, it) }
            RxMainScheduler.post {
                if (key != imageView.tag) return@post
                bitmap?.let(imageView::setImageBitmap)
                callback?.onLoaded(bitmap != null)
            }
        }
    }

    @JvmStatic
    fun clear(imageView: ImageView?) {
        imageView ?: return
        imageView.tag = null
        imageView.setImageDrawable(null)
    }

    /**
     * 清空封面内存缓存（内存压力时调用，见 LauncherActivity.onTrimMemory 方案 A）。
     * 缓存为惰性重建，清空仅导致下次加载重新解码；不主动 recycle 已交付位图。
     */
    @JvmStatic
    fun clearMemoryCache() {
        cache.evictAll()
    }

    private fun decodeSampled(context: Context, uriText: String): Bitmap? = try {
        val options = BitmapFactory.Options().apply { inJustDecodeBounds = true }
        context.contentResolver.openInputStream(Uri.parse(uriText))?.use { BitmapFactory.decodeStream(it, null, options) }
        options.inSampleSize = calculateSampleSize(options.outWidth, options.outHeight)
        options.inJustDecodeBounds = false
        context.contentResolver.openInputStream(Uri.parse(uriText))?.use { BitmapFactory.decodeStream(it, null, options) }
    } catch (error: OutOfMemoryError) {
        throw error
    } catch (_: Exception) {
        null
    }

    private fun calculateSampleSize(width: Int, height: Int): Int {
        if (width <= 0 || height <= 0) return 1
        var sample = 1
        while (width / sample > TARGET_WIDTH || height / sample > TARGET_HEIGHT) sample *= 2
        return max(1, sample)
    }
}
